[Aptos Data Client] Use non-blocking polling.

JoshLind opened the pull request:
## Motivation

This PR changes the Aptos Data Client to avoid blocking between peer polls. This addresses the issue where the poller attempts to poll a bad/slow peer and sits there until the peer responds or a timeout is hit. Given that the poller is responsible for updating the view of advertised data in the network, this causes increased latency spikes for state synchronization.

To avoid this, we don't block between polls and instead poll the peers on separate and concurrent tasks. To avoid over-polling (i.e., spawning too many tasks at once), we limit the number of outstanding polls based on an in-flight maximum for prioritized and regular peers separately.

Note: for those interested, this graph shows the number of polls currently executed per second by the nodes in devnet. The validators are currently executing ~12 a second (depending on connection churn), while the VFNs are executing anywhere from 0 to 9(!). This is a result of the fact that the peers the VFNs are polling have much higher latencies (50x-100x+) and the poller is blocking!
<img width="1296" alt="Screen Shot 2022-05-05 at 10 45 38 AM" src="https://user-images.githubusercontent.com/4578587/166949526-4bc262c4-f543-49f5-9873-f1b3c515ce54.png">

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

The new and existing unit tests pass.

## Related PRs

None, but this PR relates to: https://github.com/aptos-labs/aptos-core/issues/245

sitalkedia commented:
@JoshLind - Thanks a lot for adding this functionality. Any idea how we plan to test this end to end before re-enabling SSv2 in dev-net?

JoshLind commented:
> Any idea how we plan to test this end to end before re-enabling SSv2 in dev-net?

Through a series of painful experiments, lol 😄 There's a couple things I want to do before we even get back to re-enabling SSv2 in devnet, so it won't happen until there's a number of changes made. But, to test this specific change, the easiest way is to deploy it to a single node in devnet (e.g., one of our PFNs that is already running v2 right now), and then watch the metrics graphs (e.g., the ones I pointed at above). Ideally, the polling frequency of the PFN should match the VFNs much more closely. I'll also submit requests directly to the PFN (something we aren't doing now) to measure latency. Once we get the PFN looking good, I'll move onto the VFNs and so-forth.

sitalkedia commented on state-sync/aptos-data-client/src/aptosnet/mod.rs:
```diff
@@ -146,46 +161,60 @@ impl AptosNetDataClient {
+
+        // Fetch the last polled regular peer
+        if let Some(peer) = self.peer_states.write().oldest_polled_regular_peer() {
+            regular_peers.push(peer);
```
@JoshLind - Do we do some sort of deprioritization for slow peers or peers that are unstable?

JoshLind replied:
No. There is scoring of peers when actually requesting data, but nothing like that exists here. We can definitely add something like that in the future (and I think there's a larger reputation-style workstream that needs to happen), but I think it may be a little premature adding it now that we don't block.

sitalkedia replied:
Do we know when peers are disconnected and remove them from the peer list for future polling?

JoshLind replied:
Yes, `get_all_connected_peers()` handles this. It ensures peers are connected and that they support the storage service protocol 😄

sitalkedia replied:
@JoshLind - Maybe I am missing something here, but the `oldest_polled_regular_peer()` might return a disconnected peer as we don't seem to be updating the `regular_peer_queue` with all connected peers and removing the stale ones from the list?

JoshLind replied:
Amazing! Thank you @sitalkedia. I've updated the code to handle peer disconnect and also added tests. 

Note: The more I look at this code, the more I remember why I disagreed with the structure when it was originally ported. I think it needs to be heavily cleaned up, so I've added a comment to the code. Unfortunately, it's not high-priority yet, but if we keep needing to make "modifications" to this, I'll bite the bullet.

JoshLind commented on state-sync/aptos-data-client/src/aptosnet/mod.rs:
```diff
@@ -530,68 +555,171 @@ impl DataSummaryPoller {
+}
+
+/// Returns the value currently held in the atomic
+fn read_atomic(atomic: Arc<AtomicU64>) -> u64 {
```
Yeah, I think this entire crate needs to be split up actually. I wasn't a big fan of the layout when it was created. If it's okay, I'm going leave this to a PR in the next few days when I split things up then.

JoshLind commented on state-sync/aptos-data-client/src/aptosnet/mod.rs:
```diff
@@ -48,7 +57,10 @@ mod tests;
 // Useful constants for the Aptos Data Client
 const GLOBAL_DATA_LOG_FREQ_SECS: u64 = 5;
 const GLOBAL_DATA_METRIC_FREQ_SECS: u64 = 1;
+const MAX_NUM_IN_FLIGHT_PRIORITIZED_POLLS: u64 = 10;
```
We can make them configurable in the future, but I can't see the immediate need 😄 I think the numbers are pretty reasonable now. Thoughts?

JoshLind commented on state-sync/aptos-data-client/src/aptosnet/mod.rs:
```diff
@@ -530,68 +555,171 @@ impl DataSummaryPoller {
+            < MAX_NUM_IN_FLIGHT_REGULAR_POLLS
+        {
+            if always_poll {
+                match self.data_client.fetch_regular_peers_to_poll() {
```
Done! See: `try_fetch_regular_peers`.

JoshLind commented on state-sync/aptos-data-client/src/aptosnet/mod.rs:
```diff
@@ -146,46 +161,60 @@ impl AptosNetDataClient {
-            if let Some(peer) = self.peer_states.write().oldest_polled_regular_peer() {
-                peers_to_poll.push(peer);
+        // Mark the peers as now polled
+        for peer in &prioritized_peers {
```
Done! Added an explicit `mark_peers_as_polled` method and updated the tests.

JoshLind commented:
Actually @sitalkedia, ignore this PR for the time being. I want to make things a little cleaner and try to adopt prioritization for peers that respond more quickly. Will update.

JoshLind commented:
@sitalkedia, I've updated the PR to do the following:
1. Polling is non-blocking. We limit the number of polls by max. in-flight requests (which is configurable).
2. Polling is no longer round robin. We select randomly from the peers with the appropriate priority and without any in-flight requests. This means that peers that take longer to respond to polls are less likely to get requests. It also means that we don't over-poll peers (e.g., if the polling frequency is 100ms, but each peer takes 200ms to respond, we shouldn't be polling them more frequently than they can respond).
3. Added tests for peer connect, disconnect, reconnect and max in-flight requests.

Let me know what you think 😄

sitalkedia commented on state-sync/aptos-data-client/src/aptosnet/mod.rs:
```diff
@@ -146,46 +156,65 @@ impl AptosNetDataClient {
+    fn select_peer_to_poll(
+        &self,
+        mut peers: Vec<PeerNetworkId>,
+    ) -> Result<Option<PeerNetworkId>, Error> {
```
IIUC, previously we used to poll atleast 2 peers per poll, now we are doing just 1. Is that a concern around increasing state sync latency?

sitalkedia replied:
Also, does it makes sense to have a configurable value around number of peers we want to poll ?

JoshLind replied:
> IIUC, previously we used to poll atleast 2 peers per poll, now we are doing just 1. Is that a concern around increasing state sync latency?

IIRC (🤔) the previous world only polled 1 peer per round on average. The sampling we have now makes it 1.3 peers (i.e., 3 priority and 1 regular peers on average per 3 poll calls). This should be enough.

> Also, does it makes sense to have a configurable value around number of peers we want to poll ?

Technically, we already do. We just increase the poll frequency (which is configurable) 😄. I'm not a big of fan adding another configuration variable. Seems like overkill.

sitalkedia commented on state-sync/aptos-data-client/src/aptosnet/mod.rs:
```diff
@@ -503,95 +537,163 @@ impl DataSummaryPoller {
-                self.data_client.update_summary(peer, storage_summary);
-                self.data_client.update_global_summary_cache();
+    /// Fetches the next regular peer to poll based on the sample frequency
+    fn fetch_regular_peer(&self, always_poll: bool) -> Option<PeerNetworkId> {
```
Should we keep the sampling logic anymore? Now that you just fetch one priority peer, should we always try to get some regular peers as well?

JoshLind replied:
I think the sampling still makes sense: we want to target priority peers in the happy path and only poll regular peers infrequently or (in the unhappy path) when we have no priority peers. I don't see a benefit in always polling regular peers in the happy-path, it just seems like unnecessary work.

sitalkedia commented on state-sync/aptos-data-client/src/aptosnet/mod.rs:
```diff
@@ -503,95 +537,163 @@ impl DataSummaryPoller {
-                    update_advertised_data_metrics(global_data_summary);
+    /// Attempts to fetch the next regular peer to poll from the data client.
+    /// If an error is encountered, the error is logged and None is returned.
+    fn try_fetch_regular_peer(&self) -> Option<PeerNetworkId> {
```
nit: there is fair bit of code duplication overall between functions for fetching priority and regular peer. One way  to reduce the duplication might be to modify these functions to take `is_priority` as a flag.

JoshLind replied:
Reduced the duplication. Note: I think the readability is a little reduced but it's not bad enough to disagree with the nit 😄

sitalkedia commented on state-sync/aptos-data-client/src/aptosnet/mod.rs:
```diff
@@ -503,95 +537,163 @@ impl DataSummaryPoller {
+        data_client.update_global_summary_cache();
+
+        // Mark the in-flight poll as now complete
+        data_client.in_flight_request_complete(&peer);
```
Where are we marking the request as inflight?

JoshLind replied:
See above 😄

sitalkedia commented on state-sync/aptos-data-client/src/aptosnet/state.rs:
```diff
@@ -174,30 +172,48 @@ impl PeerStates {
-            self.prioritized_peer_queue.push_front(*peer);
+    /// Returns true iff there is an existing in-flight request
+    pub fn existing_in_flight_request(&self, peer: &PeerNetworkId) -> bool {
+        if self.is_priority_peer(peer) {
```
nit: instead of having `if .. else` for priority peer, may be a cleaner way to have a `get_in_flight_polls(is_priority)->HashSet<PeerNetworkId>` function and use that for calling `contains, insert, remove` API ?

JoshLind replied:
The issue is that you'd need two helpers: one without mutability and one with mutability. Otherwise, a call like `contains` would require mutability, which is odd. Moreover, it would then require that all lock acquisitions use the write lock, which isn't great. After having tried it, it doesn't look that great. I think it's fine to leave it as it is. If we need to add more methods which follow the pattern, then we can adopt the helpers 😄

sitalkedia commented on state-sync/aptos-data-client/src/aptosnet/mod.rs:
```diff
@@ -146,46 +156,65 @@ impl AptosNetDataClient {
+        // Select a peer at random for polling
+        let peer_to_poll = peers.choose(&mut rand::thread_rng());
+        if let Some(peer_to_poll) = peer_to_poll {
+            self.peer_states.write().new_in_flight_request(peer_to_poll);
```
Any reason to mark this as inflight over here instead of `poll_peer` function? If feasible, let's move this to the `poll_peer` function, which semantically makes more sense as we mark this as `in_flight` before the polling begins and mark it as done afterwards.

JoshLind replied:
Done.

JoshLind commented on state-sync/aptos-data-client/src/aptosnet/mod.rs:
```diff
@@ -146,46 +156,65 @@ impl AptosNetDataClient {
-        for peer in &peers_to_poll {
-            self.peer_states.write().mark_peer_as_polled(peer);
+        // Get all connected regular peers
+        let mut regular_peers = self.get_all_connected_peers()?;
```
Done. Note, I think there's a trade-off here in terms of lines of code duplication and readability, but it seems okay-ish.

sitalkedia commented on state-sync/aptos-data-client/src/aptosnet/mod.rs:
```diff
@@ -146,46 +156,70 @@ impl AptosNetDataClient {
+    /// filtering is done based on `is_priority_peer`.
+    fn select_peer_to_poll(&self, is_priority_peer: bool) -> Result<Option<PeerNetworkId>, Error> {
+        // Fetch the number of in-flight polls
+        let num_in_flight_polls = if is_priority_peer {
```
nit: this doesn't look so clean - we seem to be checking for `is_priority_peer` for almost every line of the function. Maybe the previous implementation with two separate functions was a cleaner approach? I would leave it to you to make a call. Sorry about the back and forth.

JoshLind replied:
Agreed -- I rolled back to the previous version 😄

sitalkedia reviewed (approved):
Thank you for cleaning up the code and fixing the node selection logic @JoshLind.

JoshLind commented:
Thanks @sitalkedia!

/land
